typedData/typedData.go

```diff
@@ -0,0 +1,905 @@
+package typedData
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"regexp"
+	"slices"
+	"strconv"
+	"strings"
+
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/starknet.go/utils"
+)
+
+type TypedData struct {
+	types       map[string]TypeDefinition
+	primaryType string
+	domain      Domain
+	message     map[string]any
+	revision    *revision
+}
+
+// Types returns a copy of the TypedData's type definitions map
+func (td *TypedData) Types() map[string]TypeDefinition {
+	copyMap := make(map[string]TypeDefinition, len(td.types))
+	for k, v := range td.types {
+		copyMap[k] = v
+	}
+	return copyMap
+}
+
+// PrimaryType returns the primary type name of the TypedData
+func (td *TypedData) PrimaryType() string {
+	return td.primaryType
+}
+
+// Domain returns the domain information of the TypedData
+func (td *TypedData) Domain() Domain {
+	return td.domain
+}
+
+// Message returns a copy of the TypedData's message map
+func (td *TypedData) Message() map[string]any {
+	copyMap := make(map[string]any, len(td.message))
+	for k, v := range td.message {
+		copyMap[k] = v
+	}
+	return copyMap
+}
```

Comment:
This will perform a shallow copy, which is fine for primitive types (ints, string, etc), but not fine for reference types (eg slices, structs that contains slices or pointers etc). At least if you're trying to prevent the user from modifying the td fields.

Why are we preventing the user from modifying these fields?

Reply:
I did that because when creating a Typed Data the code automatically encodes the types and assigns the result to some parameters; if the user changes any value this will not be reflected to previous computed values and so will lead to a wrong message hash. 
If the user really wants to change the message, just change the original JSON and create a new TypedData from it. As the JSON is the source of truth, there's no advantage in being able to modify a TypedData generated from it since this new TypedData no longer represents the original JSON file.

>but not fine for reference types (eg slices, structs that contains slices or pointers etc). At least if you're trying to prevent the user from modifying the td fields.

Hmm. Any good idea about that?

Reply:
We could write logic to perform a deep copy, but imo this is a bit overkill atm. I agree it's fair that the JSON is the source of truth, but imo if the user modifies the message then they should also expect the hash to be modified as well.

However, I think we should be consistent, and either perform a hard copy of each field, or allow the user to modify the types.

Reply:
Done, changed to public fields